File: 2020/236.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        return self.dfs(root, p, q)
    
    def dfs(self, root, p, q):
        if root in (None, p, q):
            return root
        left = self.dfs(root.left, p, q)
        right = self.dfs(root.right, p, q)
        return root if left and right else (left or right)

File: current_session/c++/236.cpp

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// dfs?
// 
class Solution {
public:
  TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
    return dfs(root, p, q);
  }
private:
  // TreeNode* lca;
  TreeNode* dfs(TreeNode* node, TreeNode* p, TreeNode* q) {
    if (node == nullptr || node == p || node == q) return node;
    TreeNode* left = dfs(node->left, p, q);
    TreeNode* right = dfs(node->right, p, q);
    if (left != nullptr && right != nullptr) {
      return node;
    } else if (left != nullptr) {
      return left;
    } else if (right != nullptr) {
      return right;
    } else {
      return nullptr;
    }
  }
};